client/json-rpc/src/async_client/http_client.rs

```diff
@@ -0,0 +1,435 @@
+                    return Err(Error::unexpected_invalid_response_id(resp.clone()));
+                }
+                index
+            } else {
+                return Err(Error::unexpected_invalid_response_id_type(resp.clone()));
+            }
+        } else {
+            return Err(Error::unexpected_response_id_not_found(resp.clone()));
+        };
+
+        if let Some(state) = self.last_known_state() {
+            if resp.libra_chain_id != state.chain_id {
+                return Err(Error::ChainIdMismatch(resp.clone()));
+            }
+        }
+
+        let resp_state = State::from_response(resp);
+        if !self.update_state(resp_state) {
+            return Err(Error::StaleResponseError(resp.clone()));
+        }
+
+        if let Some(ref err) = resp.error {
+            return Err(Error::JsonRpcError(err.clone()));
+        }
+        Ok(id)
+    }
+}
+
+pub struct SimpleHttpClient {
+    pub http_client: reqwest::Client,
+    pub url: reqwest::Url,
+    pub state_manager: StateManager,
+}
+
+#[async_trait]
+impl HttpClient for SimpleHttpClient {
+    async fn single_request(&self, request: &Request) -> Result<JsonRpcResponse, Error> {
+        let id = 1;
+        let rpc_resp: JsonRpcResponse = send_json_request(
+            self.http_client.clone(),
+            self.url.clone(),
+            request.to_json(id),
+        )
+        .await?;
+        self.validate(&rpc_resp, id, id)?;
+        Ok(rpc_resp)
+    }
+
+    async fn batch_request(&self, requests: &[Request]) -> Result<Vec<JsonRpcResponse>, Error> {
+        let json_requests: Vec<serde_json::Value> = requests
+            .iter()
+            .enumerate()
+            .map(|(i, r)| r.to_json(i))
+            .collect();
+
+        let rpc_resps: Vec<JsonRpcResponse> = send_json_request(
+            self.http_client.clone(),
+            self.url.clone(),
+            json![&json_requests],
+        )
+        .await?;
+        if rpc_resps.len() != requests.len() {
+            return Err(Error::unexpected_invalid_batch_response(rpc_resps));
+        }
+
+        let mut resp_maps: HashMap<usize, JsonRpcResponse> = HashMap::new();
+        for resp in rpc_resps {
+            let id = self.validate(&resp, 0, requests.len() - 1)?;
+            if resp_maps.contains_key(&id) {
+                return Err(Error::unexpected_duplicated_response_id(resp));
+            }
+            resp_maps.insert(id, resp);
+        }
+
+        json_requests
+            .iter()
+            .enumerate()
+            .map(|(i, req)| {
+                resp_maps
+                    .remove(&i)
+                    .ok_or_else(|| Error::unexpected_no_response(req.clone()))
+            })
+            .collect()
+    }
+
+    fn update_state(&self, resp_state: State) -> bool {
+        self.state_manager.update_state(resp_state)
+    }
+
+    fn last_known_state(&self) -> Option<State> {
+        self.state_manager.last_known_state()
+    }
+}
+
+pub struct LBHttpClient {
```

Comment:
nit - `LB` might give the wrong impression that this impl does a load balancing between the given urls when in fact it broadcasts.

Reply:
Updated